vpr/src/place/feasible_region_move_generator.cpp

```diff
@@ -0,0 +1,106 @@
+#include "feasible_region_move_generator.h"
+#include "globals.h"
+#include <algorithm>
+#include "math.h"
+
+e_create_move FeasibleRegionMoveGenerator::propose_move(t_pl_blocks_to_be_moved& blocks_affected, float rlim, std::vector<int>& X_coord, std::vector<int>& Y_coord, e_move_type& /*move_type*/, const t_placer_opts& placer_opts, const PlacerCriticalities* criticalities) {
+    auto& place_ctx = g_vpr_ctx.placement();
+    auto& cluster_ctx = g_vpr_ctx.clustering();
+
+    /* Pick a random block to be swapped with another random block.   */
+    // pick it from the highly critical blocks
+    if (highly_crit_pins.size() == 0) {
+        return e_create_move::ABORT; //No critical block
+    }
+    std::pair<ClusterNetId, int> crit_pin = highly_crit_pins[vtr::irand(highly_crit_pins.size() - 1)];
+    ClusterBlockId b_from = cluster_ctx.clb_nlist.net_driver_block(crit_pin.first);
+
+    if (!b_from) {
+        return e_create_move::ABORT; //No movable block found
+    }
+
+    //from block data
+    t_pl_loc from = place_ctx.block_locs[b_from].loc;
+    auto cluster_from_type = cluster_ctx.clb_nlist.block_type(b_from);
+    auto grid_from_type = g_vpr_ctx.device().grid[from.x][from.y].type;
+    VTR_ASSERT(is_tile_compatible(grid_from_type, cluster_from_type));
+
+    /* Calculate the feasible region */
+    t_pl_loc to;
+    int ipin;
+    ClusterBlockId bnum;
+    int max_x, min_x, max_y, min_y;
+
+    X_coord.clear();
+    Y_coord.clear();
+    //For critical input nodes, calculate the x&y min-max valus
+    for (ClusterPinId pin_id : cluster_ctx.clb_nlist.block_input_pins(b_from)) {
+        ClusterNetId net_id = cluster_ctx.clb_nlist.pin_net(pin_id);
+        if (cluster_ctx.clb_nlist.net_is_ignored(net_id))
+            continue;
+
+        ipin = cluster_ctx.clb_nlist.pin_net_index(pin_id);
+        if (criticalities->criticality(net_id, ipin) > placer_opts.place_crit_limit) {
+            bnum = cluster_ctx.clb_nlist.net_driver_block(net_id);
+            X_coord.push_back(place_ctx.block_locs[bnum].loc.x);
+            Y_coord.push_back(place_ctx.block_locs[bnum].loc.y);
+        }
+    }
+    if (X_coord.size() != 0) {
+        max_x = *(std::max_element(X_coord.begin(), X_coord.end()));
+        min_x = *(std::min_element(X_coord.begin(), X_coord.end()));
+        max_y = *(std::max_element(Y_coord.begin(), Y_coord.end()));
+        min_y = *(std::min_element(Y_coord.begin(), Y_coord.end()));
+    } else {
+        max_x = from.x;
+        min_x = from.x;
+        max_y = from.y;
+        min_y = from.y;
+    }
+
+    //Get the most critical output of the node
+    int xt, yt;
+    xt = 0;
+    yt = 0;
+
+    ClusterBlockId b_output = cluster_ctx.clb_nlist.net_pin_block(crit_pin.first, crit_pin.second);
+    t_pl_loc output_loc = place_ctx.block_locs[b_output].loc;
+    xt = output_loc.x;
+    yt = output_loc.y;
+
+    //determine the feasible region
+    t_bb FR_coords;
```

Comment:
Need to explain what the logic block is doing.

Reply:
Done